sling/nlp/parser/trainer/flow/builder.py

```diff
@@ -0,0 +1,248 @@
+      value = value
+    elif type(value) is int:
+      dtype = DT_INT
+      shape = []
+      value = value
+
+    # Get type and shape if missing.
+    if dtype is None: dtype = str(value.dtype)
+    if shape is None: shape = list(value.shape)
+
+    var = self.flow.var(self.varname("const"), dtype, shape)
+    var.data = value
+    return var
+
+  def opname(self, optype):
+    name = self.func.name + '/' + optype
+    if name not in self.flow.ops: return name
+    index = 1
+    while True:
+      n = name + "_" + str(index)
+      if n not in self.flow.ops: return n
+      index += 1
+
+  def varname(self, var):
+    name = self.func.name + '/' + var
+    if name not in self.flow.vars: return name
+    index = 1
+    while True:
+      n = name + "_" + str(index)
+      if n not in self.flow.vars: return n
+      index += 1
+
+  def add(self, x, y, name=None):
+    return self.op("Add", [x, y], name)
+
+  def sub(self, x, y, name=None):
+    return self.op("Sub", [x, y], name)
+
+  def mul(self, x, y, name=None):
+    return self.op("Mul", [x, y], name)
+
+  def div(self, x, y, name=None):
+    return self.op("Div", [x, y], name)
+
+  def min(self, x, y, name=None):
+    return self.op("Minimum", [x, y], name)
+
+  def max(self, x, y, name=None):
+    return self.op("Maximum", [x, y], name)
+
+  def matmul(self, x, y, name=None):
+    result = self.op("MatMul", [x, y], name)
+    result.type = x.type
+    if len(x.shape) == 2 and len(y.shape) == 2:
+      result.shape = [x.shape[0], y.shape[1]]
+    else:
+      result.shape = [0]
+    return result
+
+  def t(self, x, name=None):
+    result = self.op("Transpose", [x], name)
+    if len(x.shape) == 2:
+      result.shape = [x.shape[1], x.shape[0]]
+    else:
+      result.shape = [0]
+    return result
+
+  def log(self, x, name=None):
+    return self.op("Log", [x], name)
+
+  def exp(self, x, name=None):
+    return self.op("Exp", [x], name)
+
+  def tanh(self, x, name=None):
+    return self.op("Tanh", [x], name)
+
+  def sigmoid(self, x, name=None):
+    return self.op("Sigmoid", [x], name)
+
+  def relu(self, x, name=None):
+    return self.op("Relu", [x], name)
+
+  def sin(self, x, name=None):
+    return self.op("Sin", [x], name)
+
+  def cos(self, x, name=None):
+    return self.op("Cos", [x], name)
+
+  def square(self, x, name=None):
+    return self.op("Square", [x], name)
+
+  def neg(self, x, name=None):
+    return self.op("Neg", [x], name)
+
+  def abs(self, x, name=None):
+    return self.op("Abs", [x], name)
+
+  def sign(self, x, name=None):
+    return self.op("Sign", [x], name)
+
+  def rcp(self, x, name=None):
+    return self.op("Reciprocal", [x], name)
+
+  def identity(self, x, name=None):
+    return self.op("Identity", [x], name)
+
+  def ref(self, instance, var, name=None):
+    r = self.op("Reference", [instance], name)
+    r.producer.add_attr("var", var.name)
+    r.type = var.type
+    r.shape = var.shape
+    return r
+
+
+class Expression:
```

Comment:
We can also delete the expression class since we are not using it.

Reply:
Done